### ISSUE TITLE
[BUG_SP] Combines can't detect each other (multitool)

### ISSUE THREAD

**Issue by SilverWheelz** [Courseplay/Courseplay_FS22#776]
**READ FIRST**
https://github.com/Courseplay/Courseplay_FS22/wiki

**Steps to reproduce**
Create a multitool course with two combines.

**Actual behavior**
The combine in front had a full grain tank so it stopped and waited to be unloaded. The combine following behind it kept going and slows down as it nears the other combine but it doesn't stop and their headers collide. The convoy distance was set to 100 meters. I think they aren't detecting each other since the first combine displays left(1) on the hud while the other combine displays right(1) so both of them think they are alone in the field.

**Expected behavior**
Both combines to detect each other and stop to avoid collision or when the convoy distance has been reached.

**Screenshots**

![Screenshot (507)](https://user-images.githubusercontent.com/94125770/151706413-1733406f-ec2e-47e2-8970-6507ef7b95b1.png)

**Courseplay version**
Version used:  7.0.0.24

**Log**

[log.txt](https://github.com/Courseplay/Courseplay_FS22/files/7966232/log.txt)

**Map**
Field 5 on Deer Creek

**Vehicles/ implements involved**
2x John Deere s790 Combines w/ HD50F header by JHHG
Mod yes/no ? Yes

**Relevant script mods**
Autodrive but was inactive.

**Comment by rohne83** [comment]
it might also be a miscalculation of the distance. Was a regular issue in FS19 especially for modes like lands and spirals that the distance was not the real distance as it was calculated based on the current waypoints of respective offset course. Can imagine it's the same in FS22. Would be helpful if we could see the estimated distance anywhere.

**Comment by pull2play** [comment]
Yeah, I've noticed this too. Already mentioned it under one of the commits, as it also happens on turns. Follower will collide with leader and then overtake.

**Comment by Tensuko** [comment]
@pull2play there is only one way to report something and that @SilverWheelz have done perfectly!
Any random reports under Commits will be ignored or when I see it I delete it as it doesn't belong there.

@rohne83 you might be right, the last shown distance in the logfile is:
calculated distance from S790/14 is 20.34 m
If that is the one from the screenshot we are at that point.
We only know the progress to have the dsitance in meters is very inacurate.

**Comment by pull2play** [comment]
> @pull2play there is only one way to report something and that @SilverWheelz have done perfectly! Any random reports under Commits will be ignored or when I see it I delete it as it doesn't belong there.

I didn't report it because I was waiting on release. You make so many changes between releases it doesn't make sense to clutter up the bug list imo. Fair enough though, if you prefer I'll make bug reports as soon as I spot something :)

**Comment by Tensuko** [comment]
You can make bug reports as soon as we have the release ready but not inbetween!

**Comment by whiteeraven** [comment]
Hi,
I think, that the leader calculates the maximum distance, as stated in configuration and if the follower is too far, it will stop and wait for it to reach i.e. 100m distance from him.
Shouldn't it be that it's the follower  that should calculate, and if too near, should pause it's course?

The above is easy to replicate. Get two combines / tractors and start the leader and do not start the follower. Leader will stop after some distance and wait untill you start the follower.

At least that is what I've found while I were working on my courses.

**Comment by Tensuko** [comment]
They do both. The first check the one behind, the one behind check the front.
Somehow they think they have a higher distance then what they rly have. Thats what the log says.

**Comment by pvaiko** [comment]
Don't think the estimate is too bad, the problem is we proportionally reduce the speed with the distance, and this way we won't stop until we reach the vehicle in front of us. Just need to change the formula a bit to set the speed 0 when we reach maybe 50% of the convoy distance setting.

**Comment by SilverWheelz** [comment]
> They do both. The first check the one behind, the one behind check the front. Somehow they think they have a higher distance then what they rly have. Thats what the log says.

I think so too. If CP does indeed think that the distance between the two combines in the screenshot is 20.34m then it's overestimating. Half the header of the combine on the right plus half the header of the combime on the left should equate to the total length of one header which is around 15m since both combines were side by side. 

> the problem is we proportionally reduce the speed with the distance, and this way we won't stop until we reach the vehicle in front of us.

I did notice the following combine slow down which I like since a dynamic following speed is much better and smoother than just stop and go. The only problem is it fails to stop.

**Comment by rohne83** [comment]
Not sure if this is still valid, but in FS19 we had a very inaccurate distance calculation around the first field row waypoint for especially lands mode. Think that was because of the offset courses (most turns into the same direction, so the 'outer' vehicle had a much longer distance to the course end). A real distance of just few meters was estimated as a distance of several hundred meters by CP. With progression on the course the delta between real and estimated distance decreased until the estimate was correct for the last lane then.
Haven't tried out the convoy mode yet as the default maps have just very small fields. But another issue with lands and for large working widths also for racetracks was that a field row turn can easily have a distance of several hundred meters between last waypoint of previous row and first waypoint of next row, which the convoy mode did not take into account and resulted in the real distance growing rapidly with every turn.
Can imagine that pull request #758 is addressing either or both of those issues?

**Comment by pvaiko** [comment]
None of the two is an issue, don't think it ever was in the form you describe it, the only assumption we make currently which may not always be valid with all patterns is that the up/down part of the original course and the offset course is of the same length.

#758 is about storing waypoints in the saved course.
